biocrnpyler/compartments.py

```diff
@@ -0,0 +1,63 @@
+
+import warnings
+
+class Compartment():
+
+    """ A formal Compartment object for a Species in a CRN.
+     A Compartment must have a name. They may also have a spatial dimension (such as 2 for two-dimensional,
+     or 3 for three-dimensional) and the volume in litres. 
+    """
+
+    def __init__(self, name: str, spatial_dimensions = 3, volume=1e-6, **keywords):
+        self.name = name
+        self.spatial_dimensions = spatial_dimensions 
+        self.volume = volume 
+
+    @property
+    def name(self):
+        if self._name is None:
+            return ""
+        else:
+            return self._name
+
+    @name.setter
+    def name(self, name: str):
+        if name is None:
+            raise TypeError("Name must be a string.")
+        else:
+            no_underscore_string = name.replace("_", "")
+            if no_underscore_string.isalnum() and "__" not in name and name[len(name)-1] != "_" and not name[0].isnumeric():
+                self._name = name
+            else:
+                raise ValueError(f"name attribute {name} must consist of letters, numbers, or underscores and cannot contained double underscores or end in an underscore.")
+
+    @property
+    def spatial_dimensions(self):
+        return self._spatial_dimensions
+
+    @spatial_dimensions.setter
+    def spatial_dimensions(self, spatial_dimensions: int):
+        self._spatial_dimensions = spatial_dimensions
+
+    @property
+    def volume(self):
+        return self._volume
+
+    @volume.setter
+    def volume(self, volume: float):
+        self._volume = volume
+    
+    def __eq__(self, other):
+        """
+        Overrides the default implementation
+        Two compartments are equivalent if they have the same name, spatial dimension, and volume
+        :param other: Compartment instance
+        :return: boolean
+        """
```

Comment:
I am worried about how this will propogate through BioCRNpyler if we start using the feature. If two Species (or any subclass thereof) have different Compartments, they must be different right? It seems challenging to have to keep track of the compartment by these 3 variables. Maybe we require that each compartment have a unique name to simplify things?

Reply:
Fixing this by only looking at compartment names when comparing equality. Also, throw an error if two compartments have same name but different volume or spatial dimension.